File: backend/sanitize.py
"""
Input sanitization utilities for security.
"""
import re
import html
from typing import Optional


def sanitize_html(text: str) -> str:
    """
    Escape HTML special characters to prevent XSS attacks.
    
    Args:
        text: Input string that may contain HTML
        
    Returns:
        Sanitized string with HTML entities escaped
    """
    if not text:
        return text
    
    return html.escape(text, quote=True)


def sanitize_sql(text: str) -> str:
    """
    Basic SQL injection prevention (SQLAlchemy ORM handles this better).
    This is mainly for logging and display purposes.
    
    Args:
        text: Input string
        
    Returns:
        Sanitized string
    """
    if not text:
        return text
    
    # Remove SQL keywords (basic protection, ORM is better)
    dangerous_patterns = [
        r"(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE)\b)",
        r"(--|;|\/\*|\*\/)",
    ]
    
    result = text
    for pattern in dangerous_patterns:
        result = re.sub(pattern, "", result, flags=re.IGNORECASE)
    
    return result.strip()


def sanitize_filename(filename: str) -> str:
    """
    Sanitize filename to prevent path traversal attacks.
    
    Args:
        filename: Original filename
        
    Returns:
        Safe filename
    """
    if not filename:
        return filename
    
    # Remove path separators and special characters
    filename = re.sub(r'[/\\]', '', filename)
    filename = re.sub(r'\.\.', '', filename)
    filename = re.sub(r'[<>:"|?*]', '', filename)
    
    return filename.strip()


def validate_email(email: str) -> bool:
    """
    Validate email format.
    
    Args:
        email: Email address to validate
        
    Returns:
        True if valid email format
    """
    if not email:
        return False
    
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_username(username: str) -> bool:
    """
    Validate username format (alphanumeric, underscore, hyphen).
    
    Args:
        username: Username to validate
        
    Returns:
        True if valid username format
    """
    if not username:
        return False
    
    pattern = r'^[a-zA-Z0-9_-]{3,50}$'
    return bool(re.match(pattern, username))


def sanitize_search_query(query: str, max_length: int = 100) -> str:
    """
    Sanitize search query input.
    
    Args:
        query: Search query string
        max_length: Maximum allowed length
        
    Returns:
        Sanitized search query
    """
    if not query:
        return ""
    
    # Trim to max length
    query = query[:max_length]
    
    # Remove special characters that might be used for injection
    query = re.sub(r'[<>\'";\\]', '', query)
    
    return query.strip()


def sanitize_input(text: str, allow_html: bool = False) -> str:
    """
    General purpose input sanitization.
    
    Args:
        text: Input text to sanitize
        allow_html: Whether to allow HTML (if False, escapes HTML)
        
    Returns:
        Sanitized text
    """
    if not text:
        return text
    
    if not allow_html:
        text = sanitize_html(text)
    
    # Remove null bytes
    text = text.replace('\x00', '')
    
    return text.strip()

File: frontend/src/utils/sanitize.ts
/**
 * Sanitization utilities for frontend inputs.
 */

/**
 * Sanitize HTML input to prevent XSS attacks
 */
export const sanitizeHtml = (input: string): string => {
  if (!input) return input;
  
  const map: { [key: string]: string } = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#x27;',
    '/': '&#x2F;',
  };
  
  return input.replace(/[&<>"'/]/g, (char) => map[char]);
};

/**
 * Sanitize input for display purposes
 */
export const sanitizeInput = (input: string): string => {
  if (!input) return input;
  
  // Remove null bytes
  let sanitized = input.replace(/\x00/g, '');
  
  // Trim whitespace
  sanitized = sanitized.trim();
  
  return sanitized;
};

/**
 * Sanitize search query
 */
export const sanitizeSearchQuery = (query: string, maxLength: number = 100): string => {
  if (!query) return '';
  
  // Trim to max length
  let sanitized = query.substring(0, maxLength);
  
  // Remove special characters that might be dangerous
  sanitized = sanitized.replace(/[<>'"`;\\]/g, '');
  
  return sanitized.trim();
};

/**
 * Validate email format
 */
export const validateEmail = (email: string): boolean => {
  if (!email) return false;
  
  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  return emailRegex.test(email);
};

/**
 * Validate username format
 */
export const validateUsername = (username: string): boolean => {
  if (!username) return false;
  
  // 3-50 characters, alphanumeric, underscore, hyphen
  const usernameRegex = /^[a-zA-Z0-9_-]{3,50}$/;
  return usernameRegex.test(username);
};

/**
 * Validate password strength
 */
export const validatePasswordStrength = (password: string): {
  isValid: boolean;
  strength: number;
  requirements: {
    length: boolean;
    uppercase: boolean;
    lowercase: boolean;
    number: boolean;
    special: boolean;
  };
} => {
  const requirements = {
    length: password.length >= 8,
    uppercase: /[A-Z]/.test(password),
    lowercase: /[a-z]/.test(password),
    number: /[0-9]/.test(password),
    special: /[!@#$%^&*(),.?":{}|<>]/.test(password),
  };
  
  const metRequirements = Object.values(requirements).filter(Boolean).length;
  const strength = Math.min((metRequirements / 5) * 100, 100);
  const isValid = metRequirements >= 4; // At least 4 out of 5 requirements
  
  return { isValid, strength, requirements };
};

/**
 * Sanitize filename
 */
export const sanitizeFilename = (filename: string): string => {
  if (!filename) return filename;
  
  // Remove path separators and dangerous characters
  let sanitized = filename.replace(/[/\\]/g, '');
  sanitized = sanitized.replace(/\.\./g, '');
  sanitized = sanitized.replace(/[<>:"|?*]/g, '');
  
  return sanitized.trim();
};

/**
 * Escape regex special characters
 */
export const escapeRegex = (str: string): string => {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
};

/**
 * Check if URL is safe (prevent javascript: and data: URLs)
 */
export const isSafeUrl = (url: string): boolean => {
  if (!url) return false;
  
  const lower = url.toLowerCase().trim();
  
  // Block dangerous protocols
  const dangerousProtocols = ['javascript:', 'data:', 'vbscript:', 'file:'];
  
  for (const protocol of dangerousProtocols) {
    if (lower.startsWith(protocol)) {
      return false;
    }
  }
  
  return true;
};
